File: binding/ios/EagleProfiler.swift
//
//  Copyright 2023-2024 Picovoice Inc.
//  You may not use this file except in compliance with the license. A copy of the license is located in the "LICENSE"
//  file accompanying this source.
//  Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
//  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
//  specific language governing permissions and limitations under the License.
//

import Foundation
import PvEagle

public enum EagleProfilerEnrollFeedback {
    case AUDIO_OK
    case AUDIO_TOO_SHORT
    case UNKNOWN_SPEAKER
    case NO_VOICE_FOUND
    case QUALITY_ISSUE
}

/// EagleProfiler class for iOS Eagle text-independent speaker recognition engine.
/// It enrolls a speaker given a set of utterances and then constructs a profile for the enrolled speaker.
public class EagleProfiler: EagleBase {

    private var handle: OpaquePointer?

    private var speakerProfileSize: Int?
    private var minEnrollAudioLength: Int?

    /// Constructor.
    ///
    /// - Parameters:
    ///   - accessKey: AccessKey obtained from the Picovoice Console (https://console.picovoice.ai/)
    ///   - modelPath: Absolute path to file containing model parameters.
    ///   - device: String representation of the device (e.g., CPU or GPU) to use. If set to `best`, the most
    ///     suitable device is selected automatically. If set to `gpu`, the engine uses the first available GPU
    ///     device. To select a specific GPU device, set this argument to `gpu:${GPU_INDEX}`, where `${GPU_INDEX}`
    ///     is the index of the target GPU. If set to `cpu`, the engine will run on the CPU with the default
    ///     number of threads. To specify the number of threads, set this argument to `cpu:${NUM_THREADS}`,
    ///     where `${NUM_THREADS}` is the desired number of threads.
    /// - Throws: EagleError
    public init(
        accessKey: String,
        modelPath: String? = nil,
        device: String? = nil
    ) throws {
        super.init()

        var modelPathArg = modelPath

        if modelPath == nil {
            modelPathArg = EagleBase.resourceBundle.path(forResource: "eagle_params", ofType: "pv")
            if modelPathArg == nil {
                throw EagleIOError("Could not find default model file in app bundle.")
            }
        }

        if !FileManager().fileExists(atPath: modelPathArg!) {
            modelPathArg = try getResourcePath(modelPathArg!)
        }


        var deviceArg = device
        if device == nil {
            deviceArg = "best"
        }

        pv_set_sdk(EagleProfiler.sdk)

        var status = pv_eagle_profiler_init(
            accessKey,
            modelPathArg,
            deviceArg,
            &handle)
        if status != PV_STATUS_SUCCESS {
            let messageStack = try EagleProfiler.getMessageStack()
            throw EagleProfiler.pvStatusToEagleError(status, "EagleProfiler init failed", messageStack)
        }

        var cSpeakerProfileSizeBytes: Int32 = 0
        status = pv_eagle_profiler_export_size(
            handle,
            &cSpeakerProfileSizeBytes)
        if status != PV_STATUS_SUCCESS {
            let messageStack = try EagleProfiler.getMessageStack()
            throw EagleProfiler.pvStatusToEagleError(status, "EagleProfiler speaker_profile_size failed", messageStack)
        }
        speakerProfileSize = Int(cSpeakerProfileSizeBytes)

        var cMinAudioLengthSamples: Int32 = 0
        status = pv_eagle_profiler_enroll_min_audio_length_samples(
            handle,
            &cMinAudioLengthSamples)
        if status != PV_STATUS_SUCCESS {
            let messageStack = try EagleProfiler.getMessageStack()
            throw EagleProfiler.pvStatusToEagleError(
                status,
                "EagleProfiler enrollment_min_audio_length_sample failed",
                messageStack)
        }
        minEnrollAudioLength = Int(cMinAudioLengthSamples)
    }

    deinit {
        self.delete()
    }

    /// Releases resources acquired by EagleProfiler.
    public func delete() {
        if handle != nil {
            pv_eagle_profiler_delete(handle)
            handle = nil
        }
    }

    /// Enrolls a speaker.
    /// - Parameters:
    ///   - pcm: An array of audio samples. The audio needs to have a sample rate
    ///          equal to `.sampleRate` and be single-channel, 16-bit linearly-encoded.
    ///          In addition it must be at least `.minEnrollSamples` samples long.
    /// - Throws: EagleError
    /// - Returns: A tuple containing a the percentage of enrollment completed as a Float and
    ///            an enum representing the feedback code.
    public func enroll(pcm: [Int16]) throws -> (Float, EagleProfilerEnrollFeedback) {

        if handle == nil {
            throw EagleInvalidStateError("EagleProfiler must be initialized before enrolling")
        }

        var cEnrollError: pv_eagle_profiler_enroll_feedback_t = PV_EAGLE_PROFILER_ENROLL_FEEDBACK_AUDIO_OK
        var cPercentage: Float32 = 0
        let status = pv_eagle_profiler_enroll(
            handle,
            pcm,
            Int32(pcm.count),
            &cEnrollError,
            &cPercentage)

        if status != PV_STATUS_SUCCESS {
            let messageStack = try EagleProfiler.getMessageStack()
            throw EagleProfiler.pvStatusToEagleError(status, "EagleProfiler enroll failed", messageStack)
        }

        let enrollFeedback = pvProfilerEnrollmentErrorToEnrollFeedback(cEnrollError)
        let percentage = Float(cPercentage)

        return (percentage, enrollFeedback)
    }

    /// Exports the speaker profile. The exported profile can be used in `Eagle` or stored for later use.
    /// - Throws: EagleError
    /// - Returns: EagleProfile object that is used to perform recognition
    public func export() throws -> EagleProfile {
        if handle == nil {
            throw EagleInvalidStateError("EagleProfiler must be initialized before enrolling")
        }

        var cProfile = [UInt8](repeating: 0, count: speakerProfileSize!)
        let status = pv_eagle_profiler_export(
            handle,
            &cProfile)

        if status != PV_STATUS_SUCCESS {
            let messageStack = try EagleProfiler.getMessageStack()
            throw EagleProfiler.pvStatusToEagleError(status, "EagleProfiler export failed", messageStack)
        }

        return EagleProfile(profileBytes: cProfile)
    }

    /// Resets EagleProfiler and removes all enrollment data. It must be called before enrolling a new speaker.
    /// - Throws: EagleError
    public func reset() throws {
        if handle == nil {
            throw EagleInvalidStateError("EagleProfiler must be initialized before resetting")
        }

        let status = pv_eagle_profiler_reset(handle)

        if status != PV_STATUS_SUCCESS {
            let messageStack = try EagleProfiler.getMessageStack()
            throw EagleProfiler.pvStatusToEagleError(status, "EagleProfiler reset failed", messageStack)
        }
    }

    /// Getter for the minimum length of the input pcm required by `enroll()`.
    /// - Throws: EagleError
    /// - Returns: Minimum number of samples required for a call to `enroll()`
    public func minEnrollSamples() throws -> Int {
        if handle == nil {
            throw EagleInvalidStateError("EagleProfiler must be initialized before calling minEnrollSamples")
        }

        return minEnrollAudioLength!
    }
}

File: binding/android/Eagle/eagle/src/main/java/ai/picovoice/eagle/EagleProfiler.java
/*
    Copyright 2023 Picovoice Inc.

    You may not use this file except in compliance with the license. A copy of the license is
    located in the "LICENSE" file accompanying this source.

    Unless required by applicable law or agreed to in writing, software distributed under the
    License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
    express or implied. See the License for the specific language governing permissions and
    limitations under the License.
*/

package ai.picovoice.eagle;

import android.content.Context;
import android.content.res.Resources;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * Android binding for the profiler of the Eagle speaker recognition engine.
 * It enrolls a speaker given a set of utterances and then constructs a profile for the enrolled speaker.
 */
public class EagleProfiler {

    private static String defaultModelPath;
    private static String _sdk = "android";

    private final int minEnrollSamples;

    static {
        System.loadLibrary("pv_eagle");
    }

    private long handle;

    public static void setSdk(String sdk) {
        EagleProfiler._sdk = sdk;
    }

    /**
     * Constructor.
     *
     * @param accessKey AccessKey obtained from Picovoice Console
     * @param modelPath Absolute path to the file containing Eagle model parameters.
     * @throws EagleException if there is an error while initializing EagleProfiler.
     */
    private EagleProfiler(String accessKey, String modelPath) throws EagleException {
        EagleNative.setSdk(EagleProfiler._sdk);
        handle = EagleProfilerNative.init(accessKey, modelPath);
        minEnrollSamples = EagleProfilerNative.minEnrollSamples(handle);
    }

    /**
     * Releases resources acquired by Eagle.
     */
    public void delete() {
        if (handle != 0) {
            EagleProfilerNative.delete(handle);
            handle = 0;
        }
    }

    /**
     * Enrolls a speaker. This function should be called multiple times with different utterances of the same speaker
     * until `percentage` reaches `100.0`. Any further enrollment can be used to improve the speaker voice profile.
     * The minimum number of required samples can be obtained by calling `.getMinEnrollSamples()`.
     * The audio data used for enrollment should satisfy the following requirements:
     *     - only one speaker should be present in the audio
     *     - the speaker should be speaking in a normal voice
     *     - the audio should contain no speech from other speakers and no other sounds (e.g. music)
     *     - it should be captured in a quiet environment with no background noise
     *
     * @param pcm The audio needs to have a sample rate equal to `.getSampleRate()` and be
     *            16-bit linearly-encoded. EagleProfiler operates on single-channel audio.
     * @return The percentage of completeness of the speaker enrollment process along with the feedback code
     *         corresponding to the last enrollment attempt.
     * @throws EagleException if there is an error while enrolling speaker.
     */
    public EagleProfilerEnrollResult enroll(short[] pcm) throws EagleException {
        if (handle == 0) {
            throw new EagleInvalidStateException("Attempted to call eagle enroll after delete.");
        }

        return EagleProfilerNative.enroll(handle, pcm, pcm.length);
    }

    /**
     * Exports the speaker profile of the current session. Will raise an exception if the profile is not ready.
     *
     * @return An EagleProfile object.
     */
    public EagleProfile export() throws EagleException {
        if (handle == 0) {
            throw new EagleInvalidStateException("Attempted to call eagle profile export after delete.");
        }

        return new EagleProfile(EagleProfilerNative.export(handle));
    }

    /**
     * Resets the internal state of Eagle Profiler. It should be called before starting a new enrollment session.
     */
    public void reset() throws EagleException {
        if (handle == 0) {
            throw new EagleInvalidStateException("Attempted to call eagle reset after delete.");
        }

        EagleProfilerNative.reset(handle);
    }

    /**
     * Getter for version.
     *
     * @return Version.
     */
    public String getVersion() {
        return EagleProfilerNative.getVersion();
    }

    /**
     * Getter for audio sample rate accepted by Picovoice.
     *
     * @return Audio sample rate accepted by Picovoice.
     */
    public int getSampleRate() {
        return EagleProfilerNative.getSampleRate();
    }

    /**
     * Getter for minimum length of the input pcm required by `.enroll()`.
     *
     * @return minimum length of the input pcm.
     */
    public int getMinEnrollSamples() {
        return this.minEnrollSamples;
    }

    /**
     * Builder for creating instance of EagleProfiler.
     */
    public static class Builder {

        private String accessKey = null;
        private String modelPath = null;

        public Builder setAccessKey(String accessKey) {
            this.accessKey = accessKey;
            return this;
        }

        public Builder setModelPath(String modelPath) {
            this.modelPath = modelPath;
            return this;
        }

        private static void extractPackageResources(Context context) throws EagleIOException {
            final Resources resources = context.getResources();

            try {
                defaultModelPath = extractResource(context,
                        resources.openRawResource(R.raw.eagle_params),
                        resources.getResourceEntryName(R.raw.eagle_params) + ".pv");
            } catch (IOException ex) {
                throw new EagleIOException(ex);
            }
        }

        private static String extractResource(
                Context context,
                InputStream srcFileStream,
                String dstFilename
        ) throws IOException {
            InputStream is = new BufferedInputStream(srcFileStream, 256);
            OutputStream os = new BufferedOutputStream(context.openFileOutput(dstFilename, Context.MODE_PRIVATE), 256);
            int r;
            while ((r = is.read()) != -1) {
                os.write(r);
            }
            os.flush();

            is.close();
            os.close();
            return new File(context.getFilesDir(), dstFilename).getAbsolutePath();
        }

        /**
         * Validates properties and creates an instance of the Eagle profiler.
         *
         * @param context Android app context (for extracting Eagle resources)
         * @return An instance Eagle profiler
         * @throws EagleException if there is an error while initializing Eagle profiler.
         */
        public EagleProfiler build(Context context) throws EagleException {
            if (accessKey == null || this.accessKey.equals("")) {
                throw new EagleInvalidArgumentException("No AccessKey was provided to Eagle");
            }

            if (modelPath == null) {
                if (defaultModelPath == null) {
                    extractPackageResources(context);
                }
                modelPath = defaultModelPath;
            } else {
                File modelFile = new File(modelPath);
                String modelFilename = modelFile.getName();
                if (!modelFile.exists() && !modelFilename.equals("")) {
                    try {
                        modelPath = extractResource(context,
                                context.getAssets().open(modelPath),
                                modelFilename);
                    } catch (IOException ex) {
                        throw new EagleIOException(ex);
                    }
                }
            }

            return new EagleProfiler(accessKey, modelPath);
        }
    }

}
